packages/x-data-grid-pro/src/hooks/features/dataSource/utils.ts

```diff
@@ -0,0 +1,103 @@
+import { GridRowId } from '@mui/x-data-grid';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+
+const MAX_CONCURRENT_REQUESTS = Infinity;
+
+export const runIfServerMode = (modeProp: 'server' | 'client', fn: Function) => () => {
+  if (modeProp === 'server') {
+    fn();
+  }
+};
+
+export enum RequestStatus {
+  QUEUED,
+  PENDING,
+  SETTLED,
+  UNKNOWN,
+}
+
+/**
+ * Fetches row children from the server with option to limit the number of concurrent requests
+ * Determines the status of a request based on the enum `RequestStatus`
+ * Uses `GridRowId` to uniquely identify a request
+ */
+export class NestedDataManager {
+  private pendingRequests: Set<GridRowId> = new Set();
+
+  private queuedRequests: Set<GridRowId> = new Set();
+
+  private settledRequests: Set<GridRowId> = new Set();
+
+  private api: GridPrivateApiPro;
+
+  private maxConcurrentRequests: number;
+
+  constructor(
+    privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+    maxConcurrentRequests = MAX_CONCURRENT_REQUESTS,
+  ) {
+    this.api = privateApiRef.current;
+    this.maxConcurrentRequests = maxConcurrentRequests;
+  }
+
+  private processQueue = async () => {
+    if (this.queuedRequests.size === 0 || this.pendingRequests.size >= this.maxConcurrentRequests) {
+      return;
+    }
+    const loopLength = Math.min(
+      this.maxConcurrentRequests - this.pendingRequests.size,
+      this.queuedRequests.size,
+    );
+    if (loopLength === 0) {
+      return;
+    }
+    const fetchQueue = Array.from(this.queuedRequests);
+
+    for (let i = 0; i < loopLength; i += 1) {
+      const id = fetchQueue[i];
+      this.queuedRequests.delete(id);
+      this.pendingRequests.add(id);
+      this.api.fetchRowChildren(id);
+    }
+  };
+
+  public enqueue = async (ids: GridRowId[]) => {
+    ids.forEach((id) => {
+      this.queuedRequests.add(id);
+      this.api.unstable_dataSource.setChildrenLoading(id, true);
+    });
```

Comment:
For a large dataset calling `.setState` via `.setChildrenLoading` on each ID will be expensive. The name of the method is also not optimal, `children` refers to more than one child but the method only applies to one.

Reply:
> For a large dataset calling .setState via .setChildrenLoading on each ID will be expensive.

Correct, this could be a problem when auto-fetching so many rows. We could directly set the state here for all the rows at once. This is the only place where a large dataset is supposed to trigger `.setChildrenLoading`, so it should be fine afterward.

> children refers to more than one child but the method only applies to one.

To me, it's more about how you perceive, for example how I see the current name is "set children of row with rowId `id` to loading state", so it's referring to the children, I'd say in this perception the method _does_ apply to more than one rows.

I've also considered a few other names, but this one seemed most suitable. Here's a few of them:
**setRowLoading**: It gives the impression of only a single row being loaded, which is not the case
**setRowChildrenLoading**: It's a nice descriptive name, but to me, it's a bit longer and we could live with a shorter one

Lemme know if you prefer one of them or some other one.

Reply:
> We could directly [set](https://github.com/mui/mui-x/pull/12317/commits/35b221c8687526504b083abfcb6d63cae6069946) the state here for all the rows at once.

Sure, that sounds good. Btw would there be a memory leak with the `loading` state the same way there was with the `errors` state? Do the `false` keys stay in forever?

And did you have time to test the feature with a large dataset?

Reply:
Ok for object record.

Not sure how big our testing dataset should be, you probably have a better idea than me about that.

Reply:
It should be fine for now, since it will be more performant than the client-side tree data due to children being loaded lazily instead of being loaded in the beginning.

Even if the data size is significantly large, only the root level will load on the first render and would ideally be paginated too. We could work on optimizing if some users report the issues with large datasets.